cloth-config dependency support

MattSturgeon opened the pull request:
Following on from my initial draft PR #195, here's a subset of the changes focused on adding dependency support to cloth config.

I identified the main cause of complexity in the draft PR was auto-generating human readable descriptions of potentially complex dependencies. In hindsight, the result isn't worth the added complexity, instead it is recommended that users write their own dependency descriptions in tooltips or labels.

This PR does add a simple "Dependencies not met" line to the tooltip when `isEnabled()` is `false`. This could be removed/reformatted/reworded depending on your preference.

~~Currently, the `Dependency` may be checked multiple times per frame. If this proves performance intenseive, it may make sense to instead check once-per-tick on the main thread and store the result in an `AtomicBoolean` to reference from the render thread.~~ _implemented_

I plan to add an implementation for autoconfig in a follow up PR, based on what I have currently in the initial draft, but ideally this should be merged first - or at least the design agreed upon.

Fixes #26

MattSturgeon committed:
Enable/Display config entries using dependencies

Allows users to define dependency `Requirement`s as lambda functions, or generate them via factory methods.

A standard requirement controls whether a Config Entry is "enabled" or "disabled". When "disabled" the Config Entry is greyed-out and cannot be interacted with.

A "display" requirement controls whether a Config Entry appears _at all_ in the GUI. It's still technically present, but won't be rendered or take up any space in the list GUI.

A hidden Config Entry is also disabled automatically.

MattSturgeon commented on common/src/main/java/me/shedaniel/clothconfig2/api/ValueHolder.java:
```diff
@@ -0,0 +1,7 @@
+import org.jetbrains.annotations.Nullable;
+
+public interface ValueHolder<T> {
+    @Nullable T getValue();
```
It'd be great if this could be `NotNull`, that'd need changes in `EmptyEntry` for example though. As it stands, there's several unchecked dereferences that could cause NPEs, for example in the `Requirement` factory methods.

MattSturgeon replied:
@shedaniel are you happy to consider changes to `EmptyEntry` to allow `getValue()` to be `@NotNull`?

shedaniel replied:
I think it is a valid use case for value to be null, so no
also it could break other mods making custom config entries for button actions.

MattSturgeon replied:
OK, in that case I'll have to fix any potential NPEs before merging :+1:

MattSturgeon replied:
I've fixed the NPE issues in 688d590bfcb0ef62817795d747985167a4fbaea7 and also proposed a slightly different interface, where specific Config Entry classes can decide from themselves if their value is nullable (69d3590d2fa93476f1c9309b2ba93a398ca7eff3)

shedaniel replied:
I believe that kind of makes it more annoying to deal with, I think one single interface where it is always nullable should be fine.
I am not sure what a NotNullValueHolder would benefit, especially as the implementers should already know if they are dealing with nullable values, for example, if you are working with primitive config entries, you would know that calling getValue() will never return null, so your code is very clean anyways.

MattSturgeon replied:
Quoting JetBrains' docs:

> Overriding/implementing methods with an annotated declaration:
>
> - The `@NotNull` annotation of the parent method requires the `@NotNull` annotation for the child class method.
> - Methods with the `@Nullable` annotation in the parent method can have either `@Nullable` or `@Notaull` annotations in the child class method.

So there isn't any need for multiple interfaces; the relevant classes could simply declare the `@NotNull` annotation on their implementation.

The main justification is to make it explicit to users and IDEs that their "clean" code is actually safe without null checks. Otherwise IDEs will complain about not checking nullable values.

Adding those annotations could be part of a separate PR though

shedaniel replied:
I don't really want a spam of @NotNull annotations, I used to add them all over the place, but stopped doing it now, so you don't really see them much in newer code.
I like the idea that everything is implicitly non-null, where @Nullable is declared on nullable values. I don't really care about IDEs or even Kotlin for their strict nullability checks, it is way too much maintenance hassle for a simple thing.

MattSturgeon replied:
I've reverted back to the original `ValueHolder`. Kept the cleaner demo code, even though it technically violates `@Nullable`

MattSturgeon commented on common/src/main/resources/assets/cloth-config2/lang/en_us.json:
```diff
@@ -33,6 +33,7 @@
     "text.cloth-config.boolean.value.true": "§aYes",
     "text.cloth-config.boolean.value.false": "§cNo",
     "text.cloth-config.dropdown.value.unknown": "§cNo suggestions",
+    "text.cloth-config.disabled_tooltip": "Disabled (requirements not met)",
```
Should this have any formatting? Colour? Be reworded? :thinking: 

![image](https://github.com/shedaniel/cloth-config/assets/5046562/df63200d-a999-43ab-be03-6e373df9edfa)

Or perhaps completely remove the tooltip?

shedaniel replied:
I think this is fine

MattSturgeon commented:
@shedaniel do you think you'll get time to review this soon? I'm open to feedback and discussing the overall design if you have a different approach in mind.

shedaniel commented:
Thank you for this PR and sorry that I have completely forgotten about this, please continue to notify me if I forget in the future.
I will take a quick look later.

shedaniel commented on common/src/main/java/me/shedaniel/clothconfig2/gui/entries/TooltipListEntry.java:
```diff
@@ -55,25 +54,32 @@ public TooltipListEntry(Component fieldName, @Nullable Supplier<Optional<Compone
-        if (tooltipSupplier != null)
-            return tooltipSupplier.get();
-        return Optional.empty();
+    public Component[] getTooltip() {
```
Please don't change the method signature here (return type)

shedaniel commented on common/src/main/java/me/shedaniel/clothconfig2/gui/entries/TooltipListEntry.java:
```diff
@@ -55,25 +54,32 @@ public TooltipListEntry(Component fieldName, @Nullable Supplier<Optional<Compone
     }
     
-    public Optional<Component[]> getTooltip(int mouseX, int mouseY) {
+    public Component[] getTooltip(int mouseX, int mouseY) {
```
Please don't change the method signature here (return type)

shedaniel commented on common/src/main/java/me/shedaniel/clothconfig2/gui/widget/DynamicEntryListWidget.java:
```diff
@@ -174,7 +191,7 @@ protected final E getItemAtPosition(double mouseX, double mouseY) {
         int itemY = 0;
         int itemIndex = -1;
-        for (int i = 0; i < children().size(); i++) {
+        for (int i = 0; i < getItemCount(); i++) {
```
would change this to iterate visibleChildren so that we don't create a new visibleChildren list per children.
the `i` can be incremented separately

MattSturgeon replied:
Good point. We're actually creating _two_ `visibleChildren` lists per iteration; one in `getItemCount()` and another in `getItem()`!

MattSturgeon replied:
Optimized

shedaniel commented on common/src/main/java/me/shedaniel/clothconfig2/gui/widget/DynamicEntryListWidget.java:
```diff
@@ -456,8 +480,8 @@ protected void renderList(PoseStack matrices, int startX, int startY, int int_3,
             int itemY = startY + headerHeight;
-            for (int i = 0; i < children().size() && i < renderIndex; i++)
-                itemY += children().get(i).getItemHeight();
+            for (int i = 0; i < getItemCount() && i < renderIndex; i++)
```
Same here

MattSturgeon replied:
Optimized

shedaniel commented on common/src/main/java/me/shedaniel/clothconfig2/gui/widget/DynamicEntryListWidget.java:
```diff
@@ -497,8 +521,8 @@ protected int getRowLeft() {
         int integer = top + 4 - (int) this.getScroll() + headerHeight;
-        for (int i = 0; i < children().size() && i < index; i++)
-            integer += children().get(i).getItemHeight();
+        for (int i = 0; i < getItemCount() && i < index; i++)
```
Same here

MattSturgeon replied:
Optimized

shedaniel reviewed (changes requested):
technically these all are impl classes, but Cloth Config is so widely used now that we can't afford to break anyone's code, so we have to keep binary compat even in internal code

shedaniel commented on common/src/main/java/me/shedaniel/clothconfig2/api/AbstractConfigEntry.java:
```diff
@@ -54,6 +55,12 @@ public abstract class AbstractConfigEntry<T> extends DynamicElementListWidget.El
+    @Nullable private Requirement enableRequirement = null;
+    @Nullable private Requirement displayRequirement = null;
+    
+    private final AtomicBoolean enabled = new AtomicBoolean(true);
```
does it need to be atomic here? shouldn't really have any cross-thread issues here

MattSturgeon replied:
It's being set on the main thread and read on the render thread, but you're right that shouldn't lead to any race conditions...

Thinking about it, there could be some weird edge-cases if the value is set mid-thread; such that some render methods treat the entry as enabled while others treat it as disabled...

But Atomic doesn't solve that. Would need to lock the value for the duration of our rendering code... Which is probably overkill for single-frame visual edge-cases.

MattSturgeon replied:
We could avoid the enabled state changing mid-frame by using a frame-value cache, along the lines of:

```java
public class FrameValue<T> {
    
    // The latest value
    private T value;
    
    // The value currently in use, updated by calling sync()
    private T cached;
    
    public FrameValue(@NotNull T value) {
        this.value = this.cached = value;
    }
    
    public void set(@NotNull T value) {
        this.value = value;
    }
    
    public void sync() {
        this.cached = value;
    }
    
    public @NotNull T get() {
        return this.cached;
    }
}
```

`set()` would be called from the main thread and `get()` called multiple times per frame on the render thread.

`sync()` would be called once per frame, preferably at the start or end of the rendering. Not sure the best way to actually do that, or if its worth it to avoid single-frame race-conditions that are minor and aesthetic in nature...

MattSturgeon replied:
Switched to primitive boleans for now.

shedaniel replied:
Yeah a simple primitive boolean should do, I think tick() is actually ran on the Render Thread, both render and tick are done in Minecraft.runTick().

shedaniel commented on common/src/main/java/me/shedaniel/clothconfig2/api/Requirement.java:
```diff
@@ -0,0 +1,101 @@
+ * whose functional method is {@link #check()}.
+ */
+@FunctionalInterface
+public interface Requirement {
```
Please mark as ApiStatus.Experimental
Also, how practical is it to just use Predicate? you also get the value that could be useful.

MattSturgeon replied:
Using a `Predicate<T>` would complicate things a little.

The relevant `value` to the predicate would be the _depended-on_ Config Entry's value, but the predicate is called by the _dependent_ Config Entry.

When you consider some `Requirement.check`s might not even care about _any_ Config Entry's state, or others care about the state of multiple Config Entries, this only gets more complex...

I figured the simplest solution was to let any relevant Config Entries be referenced via closure scope when defining the `Requirement.check` lambda.

shedaniel replied:
Makes sense, thanks.

shedaniel commented on common/src/main/java/me/shedaniel/clothconfig2/api/AbstractConfigEntry.java:
```diff
@@ -92,9 +99,64 @@ public Component getDisplayedFieldName() {
+     * @see #isDisplayed()
+     * @see #tick()
+     */
+    public boolean isEnabled() {
```
Please mark all new added methods as ApiStatus.Experimental

MattSturgeon replied:
Think this is done... I've probably missed some

MattSturgeon commented:
@shedaniel I think I've addressed your points.

----

In 69d3590d2fa93476f1c9309b2ba93a398ca7eff3 I took a different approach to `ValueHolder`, splitting it into `NullableValueHolder` and `NotNullValueHolder`.

All  Config Entries implement `NullableValueHolder`, a select few config entries who know thieir value is not null override the implementation with `NotNullValueHolder`. I haven't gone through and exhaustively checked which classes can be NotNull yet, I just picked out some of the important examples.

If you don't like this approach I can revert that commit, but it will make some otherwise trivial dependencies rather verbose:

```java
builder.setRequirement(Requirement.any(
                        () -> Optional.ofNullable(dependency.getValue())
                                .map(value -> value < -70)
                                .orElseThrow(),
                        () -> Optional.ofNullable(dependency.getValue())
                                .map(value -> value > 70)
                                .orElseThrow()))
// vs
builder.setRequirement(Requirement.any(
                        () -> dependency.getValue() < -70,
                        () -> dependency.getValue() > 70))
```

----

I also went through `DynamicEntryListWidget` and optimized out a bunch of calls to `visibleChildren()` (etc) as requested.

---

I've marked the new stuff as `Experimental`, but it's getting late here so I've probably missed some bits. Did you want anything marked as `Internal` too/instead?

shedaniel commented on common/src/main/java/me/shedaniel/clothconfig2/gui/widget/DynamicEntryListWidget.java:
```diff
@@ -81,6 +83,21 @@ public DynamicEntryListWidget(Minecraft client, int width, int height, int top,
+     * 
+     * @return an unmodifiable {@link List} of visible entries
+     */
+    public List<E> visibleChildren() {
```
This will need an @Experimental

MattSturgeon replied:
Done

shedaniel commented on common/src/main/java/me/shedaniel/clothconfig2/gui/widget/DynamicEntryListWidget.java:
```diff
@@ -81,6 +83,21 @@ public DynamicEntryListWidget(Minecraft client, int width, int height, int top,
+    public List<E> visibleChildren() {
+        return this.children().stream()
+                .filter(entry -> {
+                    if (entry instanceof AbstractConfigEntry<?> configEntry)
```
Since this list widget is used across Cloth Config and also for other purposes, it would be nice if this does not reference the actual config implementations.
I would extract this to another interface, or add a isDisplayed / isVisible method to the existing Entry class.

MattSturgeon replied:
Introduced `HidableWidget` and `DisableableWidget`. Moved the implementations up to `DynamicEntryListWidget.Entry`.

shedaniel reviewed (changes requested):
Final minor nitpicks and also would need to resolve the previous issues, then I think this is good to go.
Thank you for your work so far!!

MattSturgeon commented:
Fixed your latest points. Also found and fixed a bug where a hidden boolean toggle entry was still handling click events: aa22d5e7c82dba4375986ef4440ee5de35acccf0

shedaniel commented on common/src/main/java/me/shedaniel/clothconfig2/api/DisableableWidget.java:
```diff
@@ -0,0 +1,30 @@
+     * @see HideableWidget#isDisplayed()
+     * @see TickableWidget#tick()
+     */
+    @ApiStatus.Experimental
```
You don't need these if the class is Experimental, please remove these for this class, HideableWidget and Requirement

MattSturgeon replied:
Done

shedaniel commented:
Also I am okay for explicit @Nullable, just no for explicit @NotNulls, it is fine to keep the @Nullable

shedaniel reviewed (approved):
LGTM! Thank you for everything you’ve done! I am currently on a trip, I will merge this the next weekend

MattSturgeon commented:
Great, thanks for all your help and guidance. 🎉

Would you like the PR to target a different branch or stick with `v8`?

Enjoy your trip!

shedaniel committed:
Apply license headers
